### PR TITLE
[multistage][agg] support agg with literal arguments

### PR DESCRIPTION
**Pull request by walterddr**
Just a draft but wanted to post this out for opinions CC @somandal @vvivekiyer @Jackie-Jiang @xiangfu0 

Approach Details
===
1. create a pre-rule that expands the LogicalAgg -- LogicalProject(with literal) plan and host the literals up to the LogicalAgg hints 
    - noted that at this stage we are not rewriting the project at all. only extracting the literals
    - it also extract total number of operands in case some of the agg functions requires more than 1 for placeholder purposes, it doesn't actually use them for intermediate/final stage.
2. continue normal optimization 
    - during this time, the LogicalProject can be pivoted away from the LogicalAgg, or other things could happen to completely change the plan but LogicalAgg processed during previous step will still possess the hint with literal
3. apply the Agg exchange rule to split the hint into several splits (LEAF, INTERMEDIATE, FINAL, DIRECT)
    - this RexLiterals are still carried with the hints
4. during PhysicalPlanVisitor, it will 
    - attach the literals using hints to the function context
    - attach placeholder identifiers based on how many operands it expects.

Functions added
===
- FIRSTWITHTIME/LASTWITHTIME
- DISTINCTCOUNT/DISTINCTCOUNTHLL
    - remaining DISTINCTCOUNTHLL variances are not added. can be done separately
- PERCENTILE
    - remaining PERCENTILE variances are not added. can be done separately
- stats functions (COVARPOP, COVARSAMP, VARPOP, VARSAMP, STDDEVPOP, STDDEVSAMP)
    - noted that the standard calcite notation is not yet supported, b/c putting all these functions together will result in calcite optimizing them and generates REGR_COUNT and some other non-supported AGG function

### CODE REVIEW

File: pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotAggregateLiteralAttachmentRule.java [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+        RexNode field = rexNodes.get(aggCall.getArgList().get(argIdx));
+        if (field instanceof RexLiteral) {
+          rexLiteralMap.put(new Pair<>(aggIdx, argIdx), (RexLiteral) field);
+        }
+      }
+    }
+    return rexLiteralMap;
+  }
+
+  private static String createRexLiteralHintsString(Map<Pair<Integer, Integer>, RexLiteral> rexLiterals) {
```
Comment (walterddr) [inline pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotAggregateLiteralAttachmentRule.java:114]: factor these stringify-object conversion to a util

File: pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotAggregateExchangeNodeInsertRule.java [hunk 2]
```diff
@@ -111,17 +112,18 @@ public void onMatch(RelOptRuleCall call) {
     ImmutableList<RelHint> oldHints = oldAggRel.getHints();
 
     Aggregate newAgg;
```
Comment (walterddr) [inline pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotAggregateExchangeNodeInsertRule.java:114]: note to self:
after this main `onMatch` conversion
1. we should check if all the Aggregate functions have desired literals put in place
2. we should make `AggregationFunctionType`s register optional literal checkers on specific operands

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/AggregateOperator.java [hunk 3]
```diff
@@ -234,18 +298,34 @@ private FunctionContext convertRexExpressionsToFunctionContext(RexExpression.Fun
...
+    String functionName = aggFunctionCall.getFunctionName();
+    // This should be provided by AggregationFunctionType indicating which argIdx requires literal.
+    if (functionName.equals("FIRSTWITHTIME") || functionName.equals("LASTWITHTIME")) {
+      aggArguments.add(ExpressionContext.forIdentifier("__PLACEHOLDER__"));
+      aggArguments.add(ExpressionContext.forLiteralContext(_aggCallLiteralArgsMap.get(aggIdx).get(2)));
+    }
+    if (functionName.equals("PERCENTILE")) {
+      aggArguments.add(ExpressionContext.forLiteralContext(_aggCallLiteralArgsMap.get(aggIdx).get(1)));
+    }
+  }
```
Comment (walterddr) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/AggregateOperator.java:312]: this function needs to be replaced by a util in `AggregateFunctionType` that handles each registered function separately.

**Review by xiangfu0 — APPROVED** [review]
overall lgtm

File: pinot-query-planner/src/main/java/org/apache/pinot/query/planner/logical/LiteralHintUtils.java [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    }
+    return "{" + StringUtils.join(literalStrings, ",") + "}";
+  }
+
+  public static Map<Integer, Map<Integer, Literal>> hintStringToLiteralMap(String literalString) {
+    Map<Integer, Map<Integer, Literal>> aggCallToLiteralArgsMap = new HashMap<>();
+    if (StringUtils.isNotEmpty(literalString) && !"{}".equals(literalString)) {
+      String[] literalStringArr = literalString.substring(1, literalString.length() - 1).split(",");
+      for (String literalStr : literalStringArr) {
+        String[] literalStrParts = literalStr.split("_");
```
[the remark this answers is not in the document]
Comment (walterddr) [reply pinot-query-planner/src/main/java/org/apache/pinot/query/planner/logical/LiteralHintUtils.java:55]: good point. and we shouldn't use `_` b/c some valid type name contains underscore

**Pull request merged**